src/NuGet.Core/NuGet.LibraryModel/DownloadDependency.cs

```diff
@@ -0,0 +1,108 @@
+// Copyright(c) .NET Foundation.All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+using NuGet.Shared;
+using NuGet.Versioning;
+
+namespace NuGet.LibraryModel
+{
+    public class DownloadDependency : IEquatable<DownloadDependency>, IComparable<DownloadDependency>
+    {
+        public string Name { get; }
+
+        public VersionRange VersionRange { get; }
+
+        public DownloadDependency(
+            string name,
+            VersionRange versionRange)
+        {
+            Name = name;
+            VersionRange = versionRange;
+        }
+
+        public static implicit operator LibraryRange(DownloadDependency library)
+        {
+            return new LibraryRange
+            {
+                Name = library.Name,
+                TypeConstraint = LibraryDependencyTarget.Package,
+                VersionRange = library.VersionRange
+            };
+        }
+
+        public int CompareTo(DownloadDependency other)
+        {
+
+            var compare = string.Compare(Name, other.Name, StringComparison.OrdinalIgnoreCase);
+            if (compare == 0)
+            {
+                
+                if (VersionRange == null
+                    && other.VersionRange == null)
+                {
+                    // NOOP;
+                }
+                else if (VersionRange == null)
+                {
+                    compare = -1;
+                }
+                else if (other.VersionRange == null)
+                {
+                    compare = 1;
+                }
+                else
+                {
+                    compare = VersionRange.GetHashCode().CompareTo(other.VersionRange.GetHashCode());
+                }
+            }
+            return compare;
+        }
+
+        public override string ToString()
+        {
+            var sb = new StringBuilder();
+            sb.Append(Name);
+            sb.Append(" ");
+            sb.Append(VersionRange);
+            return sb.ToString();
```

Comment:
It's not called in production code yet, it's just added for completeness, so I'm not worried about performance. 

For expanding models, I'm not sure the 1 extra object is worth the readability hit. 

Also I always thought string builder was actually faster. 
string.format uses AppendFormat, which is slower than Append.
Do you have any links to some benchmarks that suggest otherwise?

Reply:
From the book "Writing High Performance .NET Code" by Ben Watson, Kindle Web Reader says it's at "location 4092"

> For simple concatenation of a known (at compile time) quantity of strings, just use the '+' operator or the String.Concat method. This is usually more efficient than using a StringBuilder.

> Do not consider StringBuilder until the number of strings is variable and likely larger than a few dozen. The compiler will optimize simple string concatenation in a way to lessen the memory overhead.

The book also says to avoid String.Format, specifically for this case which is just a simple concatenation, but doesn't compare it to using StringBuilder for this kind of simple case.

Also, a quick google turned this up, although the the benchmark didn't compare a string builder implementation, only some string format vs concat syntaxes.

So, for performance sensitive code String.Concat seems the best, but I'm not sure how it compares to using a StringBuilder and calls to `.Append()`. Without caching a StringBuilder, I wonder if the cost of additional GC is more than the cost of AppendFormat.